### PR TITLE
fix: SSO user unable to change local account password (#3297)

### LINKED ISSUE #3297: SSO user is unable to change local user password
Checklist:

* [x] I've searched in the docs and FAQ for my answer: http://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

While SSO user with the appropriate permissions trying to change another local account password, the Argo CD requires to provide user's old password. The SSO user don't have password in Argo CD at all, so SSO user cannot manage accounts.

**To Reproduce**

* Configure SSO and ensure SSO users has admin permissions
* Create local account
* Try using `argocd account update password`

**Expected behavior**

SSO user should be able to set local account password

**Version**

```shell
v1.5.0-rc2
```

### PR DESCRIPTION
**Pull request by alexmt**
Closes #3297

### CODE REVIEW

File: server/account/account_test.go [hunk 1]
```diff
@@ -3,6 +3,7 @@ package account
 import (
```
Comment (jannfis) [inline server/account/account_test.go]: I feel we need more tests here, especially for negative outcomes. Like an SSO user trying to change password of local user, but doesn't have enough permissions to do it etc.

File: server/account/account.go [hunk 2]
```diff
@@ -52,6 +45,30 @@ func (s *Server) UpdatePassword(ctx context.Context, q *account.UpdatePasswordRe
...
+		if username == updatedUsername && rbacpolicy.IsProjectSubject(username) {
+			return nil, status.Errorf(codes.InvalidArgument, "password can only be changed for local users, not user %q", username)
+		}
+
+		err := s.sessionMgr.VerifyUsernamePassword(username, q.CurrentPassword)
+		if err != nil {
+			return nil, status.Errorf(codes.InvalidArgument, "current password does not match")
+		}
+	} else {
+		// SSO user is changing or local user password
```
Comment (jannfis) [inline server/account/account.go]: Where do we make sure that SSO user has enough permissions here?

**Review by jannfis — CHANGES_REQUESTED** [review]
Please see my comments below. Also, I'm not sure whether I fully understand the impact of this change. Where does the admin privilege for SSO user come from?

File: server/account/account.go (hunk 2 above)
Comment (alexmt) [reply server/account/account.go]: This is verified here: https://github.com/argoproj/argo-cd/pull/3298/files#diff-3dacf6e808d43913c38006b06308a2a5R41

If user is not trying to change own password we verify that user has permission to update account <account-name>.

File: server/account/account_test.go (hunk 1 above)
Comment (alexmt) [reply server/account/account_test.go]: working on it

File: server/account/account.go (hunk 2 above)
Comment (jannfis) [reply server/account/account.go]: Ah, now I see. But I think that check can be circumvented when the SSO username is the same as the local username, right? So, when there's a local user "jannfis", and someone has the SSO account "jannfis", he can change the local user's password without RBAC enforcement. Or is `sub` something different than the plain username in case of SSO?

Comment (alexmt) [reply server/account/account.go]: oh, nice catch. We need to check issuers. Working on the fix.

**Review by jessesuen — APPROVED** [review]
LGTM

File: server/account/account.go (hunk 2 above)
Comment (alexmt) [reply server/account/account.go]: implemented the fix. Now we check permissions unless the local user is trying to change own password.

File: server/account/account_test.go (hunk 1 above)
Comment (alexmt) [reply server/account/account_test.go]: added missing tests

**Comment by alexmt** [comment]
Thank you for review @jannfis . PTAL

**Review by jannfis — APPROVED** [review]
Looks much better to me now, well done! Thank you.
